### PR TITLE
fix(opentelemetry): wrap everything under Otel middleware if it Exists.

### LINKED ISSUE #3663: Bug: OpenTelemetry Middleware Doesn't Capture Exceptions Raised Before Route Handler
### Description

Recently, I integrated OpenTelemetry into my backend to improve observability. However, I've noticed a specific issue where exceptions raised before the route handler is hit aren't being logged under the current request span. This results in incomplete tracing, making it difficult to capture the full context of request lifecycles, which is crucial for accurate monitoring and debugging.

### URL to code causing the issue

_No response_

### MCVE

```python
# Your MCVE code here
```

### Steps to reproduce

```bash
1. Set up a LiteStar application with an OpenTelemetry middleware and another middleware that can raise an exception, such as a JWT middleware.
2. send request to an unknown url by the app.
3. Observe that the request span is not created.
```

### Screenshots

```bash
"![SCREENSHOT_DESCRIPTION](https://files.slack.com/files-tmb/T046DJBFEMD-F07HBK3F4TA-38f4387909/image_720.png)"
```

### Logs

```bash
Traceback (most recent call last):
  File "/Users/bella/Documents/python/litestar_backend/.venv/lib/python3.12/site-packages/litestar/_asgi/routing_trie/traversal.py", line 157, in parse_path_to_route
    node, path_parameters, path = traverse_route_map(
                                  ^^^^^^^^^^^^^^^^^^^
  File "/Users/bella/Documents/python/litestar_backend/.venv/lib/python3.12/site-packages/litestar/_asgi/routing_trie/traversal.py", line 54, in traverse_route_map
    raise NotFoundException()
litestar.exceptions.http_exceptions.NotFoundException: 404: Not Found

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/bella/Documents/python/litestar_backend/.venv/lib/python3.12/site-packages/litestar/middleware/_internal/exceptions/middleware.py", line 159, in __call__
    await self.app(scope, receive, capture_response_started)
  File "/Users/bella/Documents/python/litestar_backend/.venv/lib/python3.12/site-packages/litestar/_asgi/asgi_router.py", line 90, in __call__
    asgi_app, route_handler, scope["path"], scope["path_params"] = self.handle_routing(
                                                                   ^^^^^^^^^^^^^^^^^^^^
  File "/Users/bella/Documents/python/litestar_backend/.venv/lib/python3.12/site-packages/litestar/_asgi/asgi_router.py", line 112, in handle_routing
    return parse_path_to_route(
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/bella/Documents/python/litestar_backend/.venv/lib/python3.12/site-packages/litestar/_asgi/routing_trie/traversal.py", line 174, in parse_path_to_route
    raise NotFoundException() from e
litestar.exceptions.http_exceptions.NotFoundException: 404: Not Found
```

### Litestar Version

litestar 2.10.0

### Platform

- [X] Linux
- [X] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/3663">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/3663/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by abelkm99**
## Description

This PR addresses an issue with exception handling and span creation in the LiteStar application when using OpenTelemetry for observability. The problem occurred when exceptions were raised before the OpenTelemetry middleware was invoked, resulting in those exceptions not being captured under the current request span. This led to incomplete traces, making it difficult to monitor and debug the full lifecycle of requests.

## **Solution**

To solve this, I reordered the middleware chain to ensure that the OpenTelemetry middleware wraps the entire request lifecycle, including exception handling. The new order is as follows:

`Initial ExceptionHandler Middleware` -> `OpenTelemetry Middleware` -> `ASGIRouter` -> `Middleware n` -> `Exception Handler Middleware` -> `Route Handler`

This ensures that spans are created as early as possible in the request lifecycle and that any exceptions, regardless of where they occur, are logged within the appropriate span. **However, this will only occur if the OpenTelemetry middleware is included in the middleware stack.**

## Closes
#3663

### CODE REVIEW

**Comment by cofin** [comment]
@abelkm99 thanks for this.  I was looking to get this cleaned up only to find that there was an open PR.

**Review by cofin — APPROVED** [review]

**Comment by cofin** [comment]
@all-contributors add @abelkm99 for code, tests, and docs

**Comment by cofin** [comment]
@abelkm99 thanks again, once the tests are passing, I'll merge.

**Comment by abelkm99** [comment]
@cofin  I will fix the tests it used to pass the test but after the new merge it stoped working i will make some adjustment this afternoon. Thanks

File: litestar/app.py
Comment (provinzkraut) [inline litestar/app.py:439]: I'm hesitant to introduce more special casing here for a specific middleware. We're generally trying to move away from this.

File: litestar/app.py
Comment (provinzkraut) [inline litestar/app.py:27]: This will break for all users that don't have the `opentelemetry` package installed

**Review by provinzkraut — CHANGES_REQUESTED** [review]
I'm not so sure about the

**Comment by provinzkraut** [comment]
@abelkm99 I think the correct way of doing this would be via a plugin. Those can be referenced and looked up by name, and also alter how applications are initialised. The only reason this has to happen at this exact place right now is because of the special casing of the cors middleware, which in turn is caused by the special casing of the exception handler middleware. If we make both of those less special, that solution would be best IMO.

@litestar-org/maintainers

File: litestar/app.py
Comment (abelkm99) [reply litestar/app.py:439]: can you please elaborate more on this ?

File: litestar/app.py
Comment (abelkm99) [reply litestar/app.py:27]: I haven't thought of that.
and yes 
```
try:
    import opentelemetry  # noqa: F401
except ImportError as e:
    raise MissingDependencyException("opentelemetry") from e
```
will cause this to break

**Comment by abelkm99** [comment]
@provinzkraut  Or what if we just handle it the way the `cors_config` is handled.
so it would be like 
`self.opentelemetry_config = config.opentelemetry_config`
my thought was for previously written codes to automatically work with just an update. and that's what my implementation was about.
and also is there a plan to move the `CorsConfig` to plugin too ?

File: litestar/app.py
Comment (abelkm99) [reply litestar/app.py:439]: I am asking this cause i was thinking about different way to extract the `otel` middleware from the list of middleware's with out the need to import `OpenTelemetryInstrumentationMiddleware` since we just want to get the type and nothing model cause a string comparison  will do the job.

**Comment by cofin** [comment]
The plugin route is great! @abelkm99

@provinzkraut take a look when you have a moment.

**Comment by cofin** [comment]
>  is there a plan to move the `CorsConfig` to plugin too ?

Yeah. In general, everything is going to be moved into plugins where we can.  The plan is to also move everything in `litestar.contrib` to a plugin in `litestar.plugins`

**Comment by abelkm99** [comment]
@cofin that makes sense is it going to happen for v3 of it or the current version of litestar?

**Comment by cofin** [comment]
> @cofin that makes sense is it going to happen for v3 of it or the current version of litestar?

We'll start deprecating and moving things in 2.x, and by the time we get to Litestar 3.0, we'd like to have the `contrib` folder gone.   We've not started on this work yet though, and this is definitely something open for suggestions/help.

**Comment by provinzkraut** [comment]
Hey @abelkm99, @cofin! 
Sorry this took a while, but I've got an approach over at #3689. If you're okay with that, I'll push the changes into your branch @abelkm99.

There are three main things I've changed:
- Altered the plugin in such a way that provides a forward migration part, so it can easily work independently of the middleware in the future
- Restructure the otel middleware fix in such a way that we can upgrade it to work with middleware priorities once those are implemented (which is the *real* fix for this stuff)
- Reworked the tests to get around the hidden global state inside the otel SDK, which made testing it correctly a bit of a pain

**Comment by abelkm99** [comment]
@provinzkraut
I really appreciate the `pop_otel_middleware` logic. I hadn't considered that users might pass multiple OtelMiddleware instances. It's great to see this flexibility.
I'm also relieved that the issue with the test related to `InMemoryMetricReader` has been resolved. It was challenging when only one test at a time would pass and running several of them would fair before, which led me to skip that particular test when I noticed the `websocket` test didn't include it.
I made my best effort to address these issues independently, but I recognize that your deeper understanding of the framework's future direction allows for more comprehensive solutions.
Thank you again for reviewing my PR. Your insights have been invaluable.

**Comment by abelkm99** [comment]
I will be closing the issue and the PR now that it's resolved

**Comment by provinzkraut** [comment]
> @provinzkraut I really appreciate the `pop_otel_middleware` logic. I hadn't considered that users might pass multiple OtelMiddleware instances. It's great to see this flexibility. I'm also relieved that the issue with the test related to `InMemoryMetricReader` has been resolved. It was challenging when only one test at a time would pass and running several of them would fair before, which led me to skip that particular test when I noticed the `websocket` test didn't include it. I made my best effort to address these issues independently, but I recognize that your deeper understanding of the framework's future direction allows for more comprehensive solutions. Thank you again for reviewing my PR. Your insights have been invaluable.

Thanks for the work on this @abelkm99! I've added your commits and you as a co-author on the new PR, so the changes are attributed correctly to you :)

**Pull request closed without merging**
